dockertest/subtest.py

```diff
@@ -429,13 +458,62 @@ def try_all_stages(self, name, subsubtest):
                               self.exception_info["error_source"],
                               detail)
             exc_info = self.exception_info["exc_info"]
+            # Ignore known problems in specific docker builds
+            if self.is_known_failure(name):
+                self.final_subsubtests.add(name)
             # Treat N/A as passed, even though this may hide obscure failures
             if isinstance(detail, TestNAError):
                 self.logwarning("Treating TestNAError as PASS")
                 self.final_subsubtests.add(name)
             # cleanup() will still be called before this propagates
             raise exc_info[0], exc_info[1], exc_info[2]
 
+    def is_known_failure(self, subsubtestname):
+        """
+        Do we have a registered known failure in this subtest when running
+        on the currently-installed docker? Return True if so.
+        Side effect: log warning messages to help human debuggers.
+        """
+        # e.g. docker_cli/subtest/subsubtest
+        fullname = os.path.join(self.config_section, subsubtestname)
+        known = known_failures()
+        if fullname not in known:
+            return False
+        docker_nvr = docker_rpm()
+        if docker_nvr in known[fullname]:
+            why = known[fullname][docker_nvr]
+            self.logwarning("%s: Known failure on %s: %s",
+                            subsubtestname, docker_nvr, why)
+            return True
+
+        # This exact NVR is not known to fail. What about NV?
+        def _nv(nvr):
+            return nvr[:nvr.rfind('-')]
+        docker_nv = _nv(docker_nvr)
+
+        if docker_nv in known[fullname]:
+            why = known[fullname][docker_nv]
+            self.logwarning("%s expected to fail on all builds of %s: %s",
+                            subsubtestname, docker_nv, why)
+            return True
+
+        # No known failures for NVR or NV. What about other builds of same NV
+        # or a related one? These messages are informational only, intended
+        # as hints for a test engineer trying to understand new failures.
+        if docker_nv in [_nv(x) for x in known[fullname]]:
+            # e.g. docker is 1.12.5-6, we have an exception for 1.12.5-5
+            self.logwarning("%s is known to fail in other %s builds",
+                            subsubtestname, docker_nv)
+        elif docker_nv.count('.') > 1:
```

Comment:
Hmm, the example didn't dawn on me then.  Working in declarative languages like C has my brain stuck on "scroll up if you want to know what something is".  Anyway, if you think of it, probably wouldn't hurt to add something in case someone else is infected with declaritive-language-lookup-syndrome.  Maybe even like: ``elif docker_nv.count('.') > 1:  # See example below``